pkg/server/plugin/nodeattestor/k8s/psat/psat.go

```diff
@@ -0,0 +1,269 @@
+func NewAttestorPlugin() *AttestorPlugin {
+	return &AttestorPlugin{}
+}
+
+//AttestorPlugin is a PSAT (Projected SAT) node attestor plugin
+type AttestorPlugin struct {
+	mu     sync.RWMutex
+	config *attestorConfig
+}
+
+// ClusterConfig holds a single cluster configuration
+type ClusterConfig struct {
+	// API server public key file path.
+	// Public key is used for token validation
+	APIServerKeyFile string `hcl:"api_server_key_file"`
+
+	// Array of withelisted service accounts names
+	// Attestation is denied if comming from a service account that is not in the list
+	ServiceAccountWhitelist []string `hcl:"service_account_whitelist"`
+
+	// If true, the plugin queries k8s API Server for extra selectors
+	QueryAPIServer bool `hcl:"enable_api_server_queries"`
+
+	// Kubernetes configuration file path (only used if QueryAPIServer is enabled)
+	// Used to create a k8s client to query the API server.
+	KubeConfigFile string `hcl:"kube_config_file"`
+}
+
+// AttestorConfig contains a map of clusters that uses cluster name as key
+type AttestorConfig struct {
+	Clusters map[string]*ClusterConfig `hcl:"clusters"`
+}
+
+type clusterConfig struct {
+	keys            []crypto.PublicKey
+	serviceAccounts map[string]bool
+	queryAPIServer  bool
+	k8sClient       client.K8SClient
+}
+
+type attestorConfig struct {
+	trustDomain string
+	clusters    map[string]*clusterConfig
+}
+
+func (p *AttestorPlugin) Attest(stream nodeattestor.Attest_PluginStream) error {
+	req, err := stream.Recv()
+	if err != nil {
+		return psatError.Wrap(err)
+	}
+
+	config, err := p.getConfig()
+	if err != nil {
+		return psatError.Wrap(err)
+	}
+
+	if req.AttestedBefore {
+		return psatError.New("node has already attested")
+	}
+
+	if req.AttestationData == nil {
+		return psatError.New("missing attestation data")
+	}
+
+	if dataType := req.AttestationData.Type; dataType != pluginName {
+		return psatError.New("unexpected attestation data type %q", dataType)
+	}
+
+	if req.AttestationData.Data == nil {
+		return psatError.New("missing attestation data payload")
+	}
+
+	attestationData := new(k8s.PSATAttestationData)
+	if err := json.Unmarshal(req.AttestationData.Data, attestationData); err != nil {
+		return psatError.New("failed to unmarshal data payload: %v", err)
+	}
+
+	if attestationData.Cluster == "" {
+		return psatError.New("missing cluster in attestation data")
+	}
+
+	if attestationData.Token == "" {
+		return psatError.New("missing token in attestation data")
+	}
+
+	cluster := config.clusters[attestationData.Cluster]
+	if cluster == nil {
+		return psatError.New("not configured for cluster %q", attestationData.Cluster)
+	}
+
+	token, err := jwt.ParseSigned(attestationData.Token)
+	if err != nil {
+		return psatError.New("unable to parse token: %v", err)
+	}
+
+	claims := new(k8s.PSATClaims)
+	err = k8s.VerifyTokenSignature(cluster.keys, token, claims)
+	if err != nil {
+		return psatError.Wrap(err)
+	}
+
+	if err := claims.Validate(jwt.Expected{
```

Comment:
I wonder if we should add an expectation for Audience (and allow the expected value to be configured, defaulting to `spire-server` but allowing an empty string if people want to consciously disable audience validation).

Reply:
Looks useful! I'll add it